polys: Adding type annotations for PolyRing and PolyElement

aasim-maverick opened the pull request:
This PR adds type annotations to the code of `PolyRing` and `PolyElement` succeeding the work done in #28177.

 This PR will also in parallel focus on improving the split between implementation and interface specific logic in both the classes which will aid us in the near future while implementing the python-flint backed versions of these classes (see #28110). 

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

aasim-maverick committed:
polys: Adding type annotations for PolyRing and PolyElement

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -446,19 +451,19 @@ def term_new(self, monom, coeff) -> PolyElement[Er]:
 
     # Polynomial creation from various formats
-    def from_dict(self, element, orig_domain=None) -> PolyElement[Er]:
+    def from_dict(self, element: dict[Mon, int | Er | Expr], orig_domain: Domain[Er] | None =None) -> PolyElement[Er]:
```
The dict type is "invariant" in the value type because it is mutable. In cases like this it is better to use `Mapping` rather than dict.

aasim-maverick committed:
Using the covariant Mapping type instead of the ivariant dict(assuming we won't mutate the dict ever)

aasim-maverick committed:
Type annotations written for:
- PolyElement comparison methods
- some arithmetic methods which weren't previously annotated
- property methods
- some ring manipulations and polynomial methods

Also made a more logical split in the interface and
implementation logic of div and rem.

aasim-maverick committed:
Compatible overriding wrt IPolys and further type annotations added

aasim-maverick committed:
type annotated evaluate and subs

oscarbenjamin committed:
typ(polys): add annotations to IPolys

oscarbenjamin commented:
I've just rebased this to current master and added annotations for about half of `IPolys`. That shows what the annotations should be like there. It needed new methods `to_dup` and `from_dup`.

I don't know if you want to finish off IPolys.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1571,7 +1608,7 @@ def coeff_wrt(self, x, deg):
         monoms, coeffs = zip(*terms)
-        monoms = [m[:i] + (0,) + m[i + 1 :] for m in monoms]
+        monoms = tuple(m[:i] + (0,) + m[i + 1:] for m in monoms)
         return p.ring.from_dict(dict(zip(monoms, coeffs)))
```
This doesn't need to change from a list to a tuple. To satisfy mypy you can just give the variable a different name.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -2308,7 +2363,47 @@ def imul_num(self, c):
         return self
 
-    def _rem(self, G):
+    def _rem(self, g):
```
This is duplicating everything from `_rem_list`. Is there some advantage here in terms of efficiency compared to just doing this the obvious way:
```
def _rem(self, g):
     return self._rem_list([g])
```

oscarbenjamin committed:
typ(polys): don't import TypeAlias in Python 3.9

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1627,9 +1672,17 @@ def evaluate(self, *args, **kwargs):
+
+    @overload
+    def subs(self, x: Expr, a: Er | int) -> PolyElement[Er]:
+        ...
```
Expr is not correct here. The value gets passed to `ring.index` which expects `PolyElement[Er] | int | str`.

aasim-maverick committed:
early branching in div, reverting incorrect AI auocomplete in subs overload, removed duplication in rem helpers

aasim-maverick committed:
Further type annotations:
- type annotated pseudo div, rem, quo, exqou and respective
  helpers

- type annotated subresultants method

- reverted incorrect AI autocomplate in eval overload

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -3074,17 +3159,18 @@ def _prem(self, g, x):
             raise ZeroDivisionError("polynomial division")
 
-        q, r, dr = x, f, df
+        q, r, dr = cast('PolyElement[Er]', x), f, df
```
It is better just to rename the variable `x`.

aasim-maverick committed:
Further changes:
- Narrowed index parameter tyoe to psuedo div, rem, quo, exquo
  and subresultant helpers.

- Type annotated more PolyElement arithmetic methods.

aasim-maverick committed:
Type annotations for utility methods and split helpers

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1066,13 +1072,13 @@ def __rfloordiv__(self, other):
             raise ZeroDivisionError("polynomial division")
         elif ring.is_element(other):
-            return self._truediv(other)
+            return self._truediv(cast('Er', other))
```
This cast should not be needed and it casting to the wrong type. It should be `PolyElement[Er]`

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1257,28 +1270,28 @@ def drop(self, gen):
             return poly
 
-    def _drop_to_ground(self, gen):
+    def _drop_to_ground(self, gen: PolyElement[Er] | int | str | None) -> tuple[int, PolyRing[Er] | PolyRing[Es]]:
```
It would be better for `drop_to_ground` to convert `gen` to an `int` before calling this method.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1359,13 +1372,21 @@ def tail_degrees(self):
+    def div(self, fv: Iterable[PolyElement[Er]]) -> tuple[list[PolyElement[Er]], PolyElement[Er]]:
+        ...
+
     def div(self, fv):
```
This method should have annotations like:
```
def div(self, fv: PolyElement[Er] | Iterable[PolyElement[Er]]) -> tuple[PolyElement[Er], PolyElement[Er]] | tuple[list[PolyElement[Er]], PolyElement[Er]]
```

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1405,7 +1449,7 @@ def quo_ground(self, x):
         return self._quo_ground(x)
 
-    def extract_ground(self, g):
+    def extract_ground(self, g: PolyElement[Er]) -> tuple[Er | PolyElement[Er], PolyElement[Er], PolyElement[Er]]:
```
The first item in the return type should just be `Er` not `Er | PolyElement[Er]`.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -2508,13 +2620,24 @@ def str(self, printer, precedence, exp_pattern, mul_symbol):
+        elif i < 0:
+            return 0
+        else:
+            return self._degree(i)
+
+
+    def _degrees(self) -> tuple[int, ...]:
```
There should not be two blank lines here.

aasim-maverick commented on sympy/polys/rings.py:
```diff
@@ -785,12 +791,12 @@ def __add__(self, other: PolyElement[Er] | Er | int) -> PolyElement[Er]:
             elif (
                 isinstance(ring.domain, PolynomialRing) and ring.domain.ring == other.ring
             ):
-                return self._add_ground(other)
+                pass
             elif (
                 isinstance(other.ring.domain, PolynomialRing)
                 and other.ring.domain.ring == ring
             ):
-                return other._add_ground(self)
+                return other.__radd__(self)
```
Yes, I tried to similarly restructure these arithmetic methods in gh-28119, but I'm still unclear on how these methods should handle a specific case that SymPy supports. The problem occurs when other is a `PolyElement` from a different ring, but there's a hierarchical relationship between the two rings. Specifically, when one ring uses a `PolynomialRing` as its domain, and that domain's underlying ring matches the other polynomial's ring (or vice versa), a case for example where a `PolyRing` is used as the domain of another ring:

```python
In [1]: R, x = ring("x", ZZ) 

In [2]: R1, a = ring("a", R)

In [3]: f = x + a # should work but this enters the try branch and coercion here fails
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 f = x + a # should work but this enters the try branch and coercion here fails

TypeError: unsupported operand type(s) for +: 'PolyElement' and 'PolyElement'
 
In [4]: f = a + x # this works the other way around though, via __radd__

In [5]: f
Out[5]: a + x
```

oscarbenjamin replied:
I have pushed a simplified version in 6812d0a85ca1bc2949d5fe0c10b2ef722c71e336

oscarbenjamin committed:
polys: simplify PolyElement.__add__

oscarbenjamin committed:
Delay import of NotImplementedType for 3.9

aasim-maverick committed:
polys: simplify PolyElement.__sub__

aasim-maverick committed:
polys: Simplify PolyElement.__mul__ and refactor the simplified PolyElement.__add__ to bypass domain_new conversion

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -774,25 +789,26 @@ def __add__(self, other: PolyElement[Er] | Er | int) -> PolyElement[Er]:
-            return self.copy()
+        if self.ring.is_element(other):
+            return self._add(other)
+        res = self._try_add_ground(other)
+        if res is not NotImplemented:
+            return res
+        if isinstance(other, PolyElement):
+            domain = other.ring.domain
+            if isinstance(domain, PolynomialRing) and domain.ring.is_element(self):
+                return other._add_ground(self)
```
This doesn't work for precisely the reason I demonstrated before:
```python
In [1]: R, x = ring("x", ZZ)

In [2]: R1, a = ring("a", R)

In [3]: (x + 0*a).ring
Out[3]: Polynomial ring in x over ZZ with lex order

In [4]: (0*x + a).ring
Out[4]: Polynomial ring in a over ZZ[x] with lex order
```
We need tests that cover this case.

aasim-maverick replied:
I'm sorry, I misplaced the check by mistake, it's gonna be too late to check the case, I see. 

I will amend the commit for `__add__` and `__sub__` .

aasim-maverick committed:
polys: prevent wrong-direction domain conversions in PolyElement.__add__, PolyElement.__mul__ and PolyElement.__sub__; tests added for the same

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -846,6 +853,11 @@ def __sub__(
+        if isinstance(other, PolyElement):
+            domain = other.ring.domain
+            if isinstance(domain, PolynomialRing) and domain.ring.is_element(self):
+                return - other._sub_ground(self) # type: ignore[arg-type]
```
This should call `other._rsubground(self)` rather than `- other._sub_ground(self)`. We don't want to create an unnecessary intermediate polynomial.

oscarbenjamin commented:
> type: ignore[arg-type]

I think `cast` is better because it would not suppress from any other value on the same line.

aasim-maverick committed:
typ(polys):

- generator index directly passed to PolyElement._drop_to_ground
- type annotated PolyElement.div
- Corrected type annotations of extract_ground
- removed unnecessary overloads of PolyElement.coeff
- Corrected type annotation of extract_ground by using Sequence
  instead of Iterable
- shortened code for PolyELement._div by delegating to _div_list
- used _degree_int in _prem, _pdiv and _subresultants
- reverted unnecessary var name change while its type wasn't changing
  in _prem, _pdiv, _pquo, _pexquo, _subresultants

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1032,13 +1073,13 @@ def __rmod__(self, other):
             raise ZeroDivisionError("polynomial division")
         elif ring.is_element(other):
-            return self._floordiv(other)
+            return self._floordiv(cast('PolyElement[Er]', other))
```
Is the cast here needed?

I think that `is_element` uses `TypeIs` so it already confirms that `other` is `PolyElement[Er]`.

aasim-maverick replied:
Initially that's what I thought too, but without the cast, this error comes up with `mypy`:

> Argument 1 to "_floordiv" of "PolyElement" has incompatible type "PolyElement[Er] | Er | int"; expected "PolyElement[Er]"  [arg-type]

So I had to use `cast` here, I'm not sure why that happens despite the use of `is_element`. I don't remember exactly but perhaps this isn't the only place where this doubt struck me. 

What I thought was maybe (I really might be just wrong since I'm just speculating) since `is_element` is called on `ring` (not `other`), `mypy` might not be connecting the type guard result back to other as reliably as it would with direct `isinstance` checks. Might be a bug with `mypy`.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1653,7 +1749,7 @@ def subs(self, *args, **kwargs):
             return self._subs(subs_dict)
 
-    def prem(self, g, x=None):
+    def prem(self, g: PolyElement[Er], x: PolyElement[Er] | int | None=None) -> PolyElement[Er]:
```
When there are a type annotation for a parameter that ha a default value it reads better to have spaces around the `=` like `None = None` because `|` binds more tightly than the `=` in this case.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1485,14 +1565,14 @@ def diff(self, x):
         i = ring.index(x)
         return self._diff(i)
 
-    def trunc_ground(self, p):
+    def trunc_ground(self, p: Er) -> PolyElement[Er]:
         if self.ring.domain.is_ZZ:
             terms = []
 
             for monom, coeff in self.iterterms():
                 coeff = coeff % p
 
-                if coeff > p // 2:
+                if coeff > cast('PolyElement[Er]', p) // 2:
```
The cast here is incorrect. The type of `p` is always `Er | int` which is what the annotation should show.

This is one situation where `# type: ignore` will be needed because this method is not valid for all domains:
```python
In [1]: R = GF(3)[x]

In [2]: R.one.trunc_ground(R.domain.one)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 R.one.trunc_ground(R.domain.one)

File /stuff/current/active/sympy/sympy/polys/rings.py:1580, in PolyElement.trunc_ground(self, p)
   1578         terms.append((monom, coeff))
   1579 else:
-> 1580     terms = [(monom, coeff % p) for monom, coeff in self.iterterms()]
   1582 poly = self.new(terms)
   1583 poly.strip_zero()

TypeError: unsupported operand type(s) for %: 'flint.types.nmod.nmod' and 'flint.types.nmod.nmod'
```

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -2987,34 +3148,26 @@ def _compose(self, replacements, initial_poly):
         r = ring.zero
         term_div = self._term_div()
-        expvs = [fx.leading_expv() for fx in fv]
+        expvs = [cast('Mon', fx.leading_expv()) for fx in fv]
```
Rather than using cast here it would be better to have a method that returns the expected type:
```python
def _leading_expv(self) -> Mon:
    if not self:
        raise KeyError
    else:
        return ...

def leading_expv(self) -> Mon | None:
    try:
        return self._leading_expv()
    except KeyError:
        return None
```
Then internal code can use `_leading_expv`.

aasim-maverick commented on sympy/polys/rings.py:
```diff
@@ -2086,8 +2191,9 @@ def _pow_multinomial(self, n):
 
             for exp, (monom, coeff) in zip(multinomial, terms):
                 if exp:
-                    product_monom = monomial_mulpow(product_monom, monom, exp)
-                    product_coeff *= coeff**exp
+                    powered_monom = monomial_pow(monom, exp)
+                    product_monom = monomial_mul(product_monom, powered_monom)
+                    product_coeff *= coeff ** exp
```
I made this change a few commits back. I think I found a bug back then here. 

`monomial_mulpow` has the signature: `Callable[[Mon, int, int], Mon]` 
but the previously it was being called as 

```python
monomial_mulpow(product_monom, monom, exp)
``` 

where `monom` is `Mon` and not an `int`.

oscarbenjamin replied:
The type annotation for `monomial_mulpow` is wrong. That should be fixed rather than changing the code.

aasim-maverick committed:
polys:
- Corrected type annotation for monomial_mulpow
- avoiding extra operations in __sub__(might need further refactoring)
- removed unnecessary cast uses for Mon(updated python version locally)
- implement _leading_expv for internal code usage(returns Mon or raises KeyError)

oscarbenjamin committed:
typ(polys): annotations for Polyring/PolyElement

oscarbenjamin committed:
typ(polys): add missing PolyElement annotations

oscarbenjamin committed:
typ(polys): annotations for IPolys

oscarbenjamin committed:
type(polys) add type:ignore fro mypy 0.18

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into tmp_aasim_push

oscarbenjamin committed:
type(polys): fix annotation for dmp_lift

aasim-maverick committed:
spacing between PolyRing.clone overloads

oscarbenjamin committed:
fmt(polys): run ruff format on rings.py
